### PR TITLE
Add SNI to TCPRouteMatch

### PR DESCRIPTION
**Pull request by JeremyOT**
Per discussion at 7/23 community meeting, adding SNI to TCPRoute

### CODE REVIEW

**Comment by robscott** [comment]
Thanks!

/lgtm

**Comment by jpeach** [comment]
I think that this is a pretty simple change that will work.

However, a slightly different way to address this is to follow the same structure as `HTTPRoute`. In this case, the hostname is the root of the spec structure rather than a subsequent match criteria.

In the Go types, that looks like this diff:
```Go
diff --git apis/v1alpha1/tcproute_types.go apis/v1alpha1/tcproute_types.go
index 802bd5f..0c5a5c3 100644
--- apis/v1alpha1/tcproute_types.go
+++ apis/v1alpha1/tcproute_types.go
@@ -21,8 +21,14 @@ import (
 
 // TCPRouteSpec defines the desired state of TCPRoute
 type TCPRouteSpec struct {
-	// Rules are a list of TCP matchers and actions.
-	Rules []TCPRouteRule `json:"rules" protobuf:"bytes,1,rep,name=rules"`
+	// Hosts is a list of Host definitions.
+	Hosts []TCPRouteHost `json:"hosts,omitempty" protobuf:"bytes,1,rep,name=hosts"`
+
+	// Default is the default host to use. Default.Hostnames must
+	// be an empty list.
+	//
+	// +optional
+	Default *TCPRouteHost `json:"default" protobuf:"bytes,2,opt,name=default"`
 }
 
 // TCPRouteStatus defines the observed state of TCPRoute
@@ -42,6 +48,20 @@ type TCPRoute struct {
 	Status TCPRouteStatus `json:"status,omitempty" protobuf:"bytes,4,opt,name=status"`
 }
 
+type TCPRouteHost struct {
+	// Hostnames is a list of TLS Server Name Indication ([RFC6066](https://tools.ietf.org/html/rfc6066))
+	// names that match this route.
+	//
+	// +optional
+	Hostnames []string `json:"hostnames,omitempty" protobuf:"bytes,1,opt,name=hostnames"`
+
+	// Rules are a list of HTTP matchers, filters and actions.
+	Rules []TCPRouteRule `json:"rules" protobuf:"bytes,2,rep,name=rules"`
+
+	// +optional
+	ExtensionRef *RouteHostExtensionObjectReference `json:"extensionRef" protobuf:"bytes,3,opt,name=extensionRef"`
+}
+
 // TCPRouteRule is the configuration for a given rule.
 type TCPRouteRule struct {
 	// Match defines which connections match this rule.

```

Advantages of this approach are that it's consistent with `HTTPRoute`, and there's an additional opportunity to add an extension ref for each SNI name. Disadvantages are that the binding to the SNI requirement is less obvious, which could make the API harder to reason about and prone to misunderstanding. Maybe the SNI name ends up not being used a lot, in which case having to specify TCP routes through the `Default` field would end up as an awkward indirection.

**Comment by JeremyOT** [comment]
@jpeach this is a great point and definitely worth a discussion

File: apis/v1alpha1/tcproute_types.go [hunk 1]
```diff
@@ -97,6 +97,11 @@ type TCPRouteMatch struct {
 	//
 	// +optional
 	ExtensionRef *RouteMatchExtensionObjectReference `json:"extensionRef" protobuf:"bytes,1,opt,name=extensionRef"`
+	// SNI is the name to match based on TLS Server Name Indication
+	// ([RFC6066](https://tools.ietf.org/html/rfc6066)).
+	//
+	// +optional
+	SNI *string `json:"sni" protobuf:"bytes,2,opt,name=sni"`
```
Comment (danehans) [inline apis/v1alpha1/tcproute_types.go:104]: IMO SNI should not be part of a TCPRoute. Routes should contain matchers that are applicable to the protocol. For example, TCPRoute should match based on TCP header information.

Comment (danehans) [reply apis/v1alpha1/tcproute_types.go:104]: Maybe ^ is a use case for a TLSRoute that contains a SNI matcher and forwards to a TCPRoute.

**Comment by robscott** [comment]
I think the consensus seems to be that SNI matching does not belong on TCPRoute and instead could make more sense on a separate kind of route in the future. I'm going to close this but feel free to reopen if you'd like to explore this further.

/close

**Pull request closed without merging**